lex/lexer.ml

```diff
@@ -0,0 +1,306 @@
+  (* Return the next {n} characters in the buffer *)
+  let p = lexbuf.Lexing.lex_curr_p in
+  let c_pos = p.Lexing.pos_cnum in
+  
+  (* if there is not enough characters in the buffer, fill it again *)
+  if lexbuf.Lexing.lex_buffer_len < n + c_pos
+  then begin
+    let curr_len = lexbuf.Lexing.lex_buffer_len in
+    lexbuf.Lexing.refill_buff lexbuf;
+    let new_len = lexbuf.Lexing.lex_buffer_len in
+    if curr_len <> new_len
+    then get_next_buff lexbuf n
+    else if c_pos = curr_len
+        then Bytes.of_string ""
+        else Bytes.sub lexbuf.Lexing.lex_buffer c_pos (curr_len - c_pos)
+  end else
+    Bytes.sub lexbuf.Lexing.lex_buffer c_pos n
+
+let advance lexbuf n =
+  (* Advance the buffer by {n} characters *)
+  let p = lexbuf.Lexing.lex_curr_p in
+  let cnum, lnum, bol = ref p.Lexing.pos_cnum, ref p.Lexing.pos_lnum, ref p.Lexing.pos_bol in
+  let buff = get_next_buff lexbuf n in
+  for i = 0 to (Bytes.length buff - 1) do
+    if (Bytes.get buff i) = '\010'
+    then begin incr lnum; bol := !cnum end;
+    incr cnum
+  done;
+  lexbuf.Lexing.lex_curr_p <- { p with
+        Lexing.pos_cnum = !cnum;
+        Lexing.pos_lnum = !lnum;
+        Lexing.pos_bol = !bol;
+  }
+
+let lex_str lexbuf str =
+  let buf = get_next_buff lexbuf (String.length str) in
+
+  for i = 0 to (String.length str - 1) do
+    if (Bytes.length buf) <= i then raise Bad_rule;
+    if (String.get str i) <> (Bytes.get buf i)
+    then raise Bad_rule;
+  done;
+
+  advance lexbuf (String.length str)
+
+let rec lex_action lexbuf =
+  let buff = get_next_buff lexbuf 200 in
+  let i = ref 0 in
+  while (!brace_depth > 0) && (!i < (Bytes.length buff)) do
+    begin match (Bytes.get buff !i) with
+    | '{' -> incr brace_depth
+    | '}' -> decr brace_depth
+    | _ -> ()
+    end;
+    incr i
+  done;
+  advance lexbuf !i;
+  if !brace_depth <> 0
+  then lex_action lexbuf
+;;
+
+let rec lex_string lexbuf =
+  let buff = get_next_buff lexbuf 200 in
+  let i = ref 0 in
+  let continue = ref true in
+  let str = ref [] in
+  while (!continue = true) && (!i < Bytes.length buff) do
+    begin match Bytes.get buff !i with
+    | '\\' -> begin match Bytes.get buff (!i+1) with
+      | '\\' -> str := '\\' :: !str
+      | '\'' -> str := '\'' :: !str
+      | '"' -> str := '"' :: !str
+      | 'n' -> str := '\010' :: !str
+      | 't' -> str := '\009' :: !str
+      | 'b' -> str := '\008' :: !str
+      | 'r' -> str := '\013' :: !str
+      | ' ' -> str := ' ' :: !str
+      (* We skip many cases that do not happen in lexer.mll, such as \\[0-9]{3},
+       * or \\ before a new line. *)
+      | _ -> raise Bad_rule
+      end; incr i
+    | '"' -> continue := false
+    | c -> str := c :: !str
+    end;
+    incr i
+  done;
+  advance lexbuf !i;
+  let buf = Buffer.create (List.length !str) in
+  List.iter (Buffer.add_char buf) !str;
+  if !continue = true
+  then (Buffer.contents buf)::(lex_string lexbuf)
+  else (Buffer.contents buf)::[];;
```

Comment:
I find it difficult to understand how this function works. If I understand correctly, the characters within the string literal are read in order, each pushed to the front of `str`. By the end of the loop, `str` contains the string literal *reversed* (last character first in the list). Then you do a `List.iter (Buffer.add_char buf)`, which adds the character in `str` order, that is in reverse order (last character added first). If I understand correctly, the resulting `Buffer.contents buf` string should be the content of the string literal, *reversed*. This sounds wrong / not what we were looking for. Am I missing something?

Note: instead of adding to `str` and then adding to a buffer after the fact, why not create a buffer at the start and use `Buffer.add_char` within the loop?